### ISSUE TITLE
Promise reference in core?

### ISSUE THREAD

**Issue by okwolf** [hyperapp/hyperapp#451]
Take a look at this line currently in master: https://github.com/hyperapp/hyperapp/blob/ffe9209cd2a5ce9927c0dfb55830c4577459d2c4/src/index.js#L65

What's the purpose of `&& !result.then`?

Does Hyperapp really care if the result is a `Promise`?

All the tests still pass with this removed, including one that actually returns a `Promise`: https://github.com/hyperapp/hyperapp/blob/ffe9209cd2a5ce9927c0dfb55830c4577459d2c4/test/actions.test.js#L75-L100

What am I missing? 🤔

**Comment by JorgeBucaran** [comment]
@okwolf The purpose of that is to skip patch if an action returns a promise.

**Comment by okwolf** [comment]
> @JorgeBucaran The purpose of that is to skip patch if an action returns a promise.

OK, but what's the reasoning behind doing this for `Promise`? I thought I heard someone say we specifically wanted **not** to add special support for them into core. Why have special snowflake code for such things? ❄️  Why not also skip patch with the ids returned by [`setTimeout`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Return_value) or [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame#Return_value) while we're at it? Or any other number of side-effectful APIs that aren't `Promise`s.

**Comment by JorgeBucaran** [comment]
@okwolf I'll explain it with an example:

```
const actions = {
  fetchWallop: (state, { makeMoney }) => async coinUrl => {
    const gold = await fetch(coinUrl).then(parse)
    makeMoney(gold)
  }
}
```

- Currently, we call patch only once in this code.
- Without `!result.then` we'd call patch twice.

**Comment by JorgeBucaran** [comment]
> @okwolf Why not also skip patch with the ids returned by setTimeout or requestAnimationFrame while we're at it? Or any other number of side-effectful APIs that aren't Promises.

This is a good idea. I'll think about it.

**Comment by Swizz** [comment]
I think for side effect you strongly need to use `() => { ... }` that mean you will return nothing. 

But this case about Promise is particular. 

First using only the then presence is sufficient regarding the PromiseA+ spec so we are not supporting Promise officially in core as we are aware about all Promise implementations without the need of any polyfill.

Then, returning the Promise without executing the repaint allow you to reuse your actions resolving somewhere else in you code. 

You will be able to await your actions to resolve.

```js
await actions.asyncAction()
```

**Comment by okwolf** [comment]
> @JorgeBucaran This is a good idea. I'll think about it.

My point was that we should have less of this, not more. 😂

**Comment by okwolf** [comment]
> @Swizz I think for side effect you strongly need to use `() => { ... }` that mean you will return nothing.

I agree, why do we need anything more special than this to have side effects skip `patch`?

**Comment by Swizz** [comment]
Allowing user to reuse Promise resolving to await async actions in sequence is really helpful.

**Comment by lukejacksonn** [comment]
> What's the purpose of `&& !result.then`

To skip an update if what was returned from the action was a promise of a partial.

> Does Hyperapp really care if the result is a Promise?

Yes, because a promise indicates a future, not the value itself. A promise returns an object. So if we didn't make this check then it would be mistaken for a partial  and merged with state. Although I haven't played with the idea it does seem a bit funky.

> All the tests still pass with this removed, including one that actually returns a Promise:

This is probably because of the point above? I haven't looked into the tests of `0.16.0` yet.


A question for you.. what harm is this causing you, why should we remove this bit of code? As @Swizz says:

> Allowing user to reuse Promise resolving to await async actions in sequence is really helpful.

**Comment by JorgeBucaran** [comment]
> @okwolf My point was that we should have less of this, not more. 😂

That's why I didn't add any of those checks. I didn't miss them nor I added promise support accidentally.

**Comment by JorgeBucaran** [comment]
I am going to close here because the question has been answered, if there's another question or proposal that you would like to make, please create a new issue! 👍
